File: Java/Lesson024_WordSearchIi.java
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 024 -- Word Search II
// Category   : Tries
// Difficulty : Hard
// Study Plan : Day 12
// =============================================================
//
// QUESTION:
//   Given an m x n board of characters and a list of strings words, return
//   all words on the board. Each word must be constructed from letters of
//   sequentially adjacent cells (horizontal/vertical). The same cell may not
//   be used more than once in a word.
//
//   Example:
//     board=[["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]]
//     words=["oath","pea","eat","rain"]
//     Output: ["oath","eat"]
// =============================================================

import java.util.*;
public class Lesson024_WordSearchIi {
    static class Node { Map<Character,Node> ch = new HashMap<>(); String word; }
    public List<String> findWords(char[][] board, String[] words) {
        Node root = new Node();
        for (String w : words) { Node n = root; for (char c : w.toCharArray()) n = n.ch.computeIfAbsent(c, k -> new Node()); n.word = w; }
        Set<String> out = new HashSet<>();
        for (int r = 0; r < board.length; r++)
            for (int c = 0; c < board[0].length; c++) dfs(board, r, c, root, out);
        return new ArrayList<>(out);
    }
    void dfs(char[][] b, int r, int c, Node n, Set<String> out) {
        char ch = b[r][c];
        Node nxt = n.ch.get(ch); if (nxt == null) return;
        if (nxt.word != null) out.add(nxt.word);
        b[r][c] = '*';
        int[][] D = {{1,0},{-1,0},{0,1},{0,-1}};
        for (int[] d : D) {
            int nr=r+d[0], nc=c+d[1];
            if (nr>=0&&nr<b.length&&nc>=0&&nc<b[0].length&&b[nr][nc]!='*') dfs(b,nr,nc,nxt,out);
        }
        b[r][c] = ch;
    }
    public static void main(String[] args) {
        char[][] b = {{'o','a','a','n'},{'e','t','a','e'},{'i','h','k','r'},{'i','f','l','v'}};
        List<String> r = new Lesson024_WordSearchIi().findWords(b, new String[]{"oath","pea","eat","rain"});
        Collections.sort(r); System.out.println(r);
    }
}

File: R/Lesson_024_word_search_ii.R
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 024 -- Word Search II
# Category   : Tries
# Difficulty : Hard
# Study Plan : Day 12
# =============================================================
#
# QUESTION:
#   Given an m x n board of characters and a list of strings words, return
#   all words on the board. Each word must be constructed from letters of
#   sequentially adjacent cells (horizontal/vertical). The same cell may not
#   be used more than once in a word.
#
#   Example:
#     board=[["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]]
#     words=["oath","pea","eat","rain"]
#     Output: ["oath","eat"]
# =============================================================

# Recursive DFS with a hash set for words.
findWords <- function(board, words) {
    R <- nrow(board); C <- ncol(board)
    out <- character(0)
    dict <- new.env(hash=TRUE); for (w in words) assign(w, TRUE, envir=dict)
    prefix_set <- new.env(hash=TRUE)
    for (w in words) for (i in seq_len(nchar(w))) assign(substr(w,1,i), TRUE, envir=prefix_set)
    visited <- matrix(FALSE, R, C)
    dfs <- function(r, c, path) {
        if (r<1 || r>R || c<1 || c>C || visited[r,c]) return()
        np <- paste0(path, board[r,c])
        if (!exists(np, envir=prefix_set, inherits=FALSE)) return()
        if (exists(np, envir=dict, inherits=FALSE)) out[[length(out)+1]] <<- np
        visited[r,c] <<- TRUE
        for (d in list(c(1,0),c(-1,0),c(0,1),c(0,-1))) dfs(r+d[1], c+d[2], np)
        visited[r,c] <<- FALSE
    }
    for (r in seq_len(R)) for (c in seq_len(C)) dfs(r, c, "")
    sort(unique(out))
}
b <- matrix(c("o","a","a","n",
              "e","t","a","e",
              "i","h","k","r",
              "i","f","l","v"), nrow=4, byrow=TRUE)
print(findWords(b, c("oath","pea","eat","rain")))
